File: solutions/0198-house-robber/solution.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        dp = [-1]*n
        dp[0] = nums[0]
        def f(i):
            if i <0:
                return 0
            if i == 0:
                return nums[0]
            if dp[i] != -1:
                return dp[i]
            dp [i] = max((nums[i]+f(i-2)),f(i-1))
            return dp[i]
        f(n-1)
        return dp[n-1]

        

File: solutions/0198-house-robber/solution.java
class Solution {

    int[] dp;
    int[] nums;

    public int f(int i) {
        if (i == 0) return nums[0];
        if (i == 1) return Math.max(nums[0], nums[1]);

        if (dp[i] != -1) return dp[i];

        dp[i] = Math.max(f(i - 1), f(i - 2) + nums[i]);
        return dp[i];
    }

    public int rob(int[] nums) {
        int n = nums.length;

        if (n == 0) return 0;
        if (n == 1) return nums[0];

        this.nums = nums;
        this.dp = new int[n];
        Arrays.fill(dp, -1);

        return f(n - 1);
    }
}
